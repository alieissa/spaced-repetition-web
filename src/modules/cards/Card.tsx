/** @format */

import { MouseEventHandler, useState } from 'react'
import 'semantic-ui-css/semantic.min.css'
import { Button, Icon, Segment } from 'semantic-ui-react'
import 'src/App.css'
import { Settings } from 'src/components'
import { styles } from 'src/styles'
import { SubmittableCardForm } from '.'
import { NCards } from './cards.types'

/**
 * Displays the card (question) content or edit form when user selects edit. User can
 * also delete or update settings of a card (question) in this component
 */
export default function Card(props: NCards.Card) {
  const [editing, setEditing] = useState(false)
  return (
    <Segment className="justify-space-between align-center">
      {editing ? (
        <SubmittableCardForm
          {...props}
          onSubmit={(data) => console.log('question form data', data)}
          onCancel={() => setEditing(false)}
        />
      ) : (
        <Question
          question={props.question}
          onEdit={() => setEditing(true)}
          onDelete={() => console.log('on delete')}
          onSubmitSettings={() => console.log('submitSettings')}
        />
      )}
    </Segment>
  )
}

interface QuestionProps {
  question: NCards.Card['question']
  onEdit: MouseEventHandler
  onDelete: MouseEventHandler
  // TODO add type
  onSubmitSettings: any
}
/**
 * Displays the content of the question and the delete, edit and settings buttons
 * from which user can delete, open edit form, or open settings menu
 */
function Question(props: QuestionProps) {
  const [open, setOpen] = useState(false)
  return (
    <>
      <span className="ellipsis">{props.question}</span>
      <span>
        <Button
          style={styles.bgWhite}
          icon={<Icon name="pencil" />}
          onClick={props.onEdit}
        />
        <Button
          style={styles.bgWhite}
          icon={<Icon name="x" color="red" />}
          onClick={props.onDelete}
        />
        <Settings
          id="dummyId2"
          deckId="dummyDeck1Id"
          decks={[]}
          open={open}
          easiness={1}
          quality={1}
          interval={1}
          trigger={
            <Button
              style={styles.bgWhite}
              icon={<Icon name="setting" />}
              onClick={() => setOpen(true)}
            />
          }
          onCancel={() => setOpen(false)}
          onSave={props.onSubmitSettings}
        />
      </span>
    </>
  )
}
