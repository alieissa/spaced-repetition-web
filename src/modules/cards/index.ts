/** @format */

export { default as Card } from './Card'
export { default as CardForm } from './CardForm'
export { default as TestCard } from './TestCard'
export * from './cards.hooks'
export { default as cardsReducer } from './cards.reducer'
export * from './cards.types'

