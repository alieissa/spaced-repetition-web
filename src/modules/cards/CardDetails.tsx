/** @format */

import { useState } from 'react'
import { Button } from 'semantic-ui-react'
import { NCards } from './cards.types'

/**
 * This component contains the card details and a button that is used to display
 * the card edit form.
 */
export default function CardDetailsModal(props: NCards.Card) {
  const [isCardDetailsEditFormOpen, setIsCardDetailsEditFormOpen] =
    useState(false)

  return (
    <>
      <Button
        data-testid="card-details-edit-btn"
        onClick={() => setIsCardDetailsEditFormOpen(true)}
      />

      {isCardDetailsEditFormOpen && <CardDetailsForm />}
      <CardDetailsView {...props} />
    </>
  )
}

function CardDetailsForm() {
  return <div data-testid="card-details-form">dummy edit form</div>
}

function CardDetailsView(props: NCards.Card) {
  const [areAnswersVisible, setAreAnswersVisible] = useState(false)
  return (
    <div>
      {props.question}
      <Button
        data-testid="view-answers-btn"
        onClick={() => setAreAnswersVisible(!areAnswersVisible)}
      />
      {areAnswersVisible && (
        <div>
          {props.answers.map((answer, index) => {
            return <div data-testid={`answers-${index}`}>{answer.content}</div>
          })}
        </div>
      )}
    </div>
  )
}
