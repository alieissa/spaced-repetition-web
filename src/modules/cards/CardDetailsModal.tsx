/** @format */

import { useFormik } from 'formik'
import _ from 'lodash'
import { useEffect, useState } from 'react'
import { useNavigate, useParams } from 'react-router-dom'
import { Form, Icon, List, ListItem } from 'semantic-ui-react'
import 'src/App.css'
import {
  SPButton,
  SPModal,
  SPModalContent,
  SPModalHeader,
  SPText,
} from 'src/components'
import { styles } from 'src/styles'
import * as Yup from 'yup'
import { NAnswers } from '../answers'
import { useCardDetails } from './cards.hooks'
import { NCards } from './cards.types'

/**
 * This component contains the card details modal. The modal displays the
 * details of a card and an edit form when user clicks on edit.
 *
 * The modal is not controlled by any parent component, it is displayed when
 * the route /decks/:deckId/cards/:cardId is hit and it controls its state
 * completely
 */
export default function CardDetailsModal() {
  const params = useParams()
  const navigate = useNavigate()
  const [isModalOpen, setIsModalOpen] = useState(false)
  const [isEditing, setIsEditing] = useState(false)
  const [loadCardStatus, card, loadCard] = useCardDetails(
    params.deckId!,
    params.cardId!,
  )

  const handleCloseModal = () => {
    setIsModalOpen(false)
    setIsEditing(false)
    navigate(`/decks/${params.deckId}`)
  }

  useEffect(() => {
    loadCard()
  }, [params.cardId])

  useEffect(() => {
    if (loadCardStatus.type !== 'Success') {
      return
    }

    setIsModalOpen(true)
  }, [loadCardStatus.type])

  return (
    <SPModal
      data-testid="card-details-modal"
      open={isModalOpen}
      onClose={handleCloseModal}
    >
      {isEditing ? (
        <CardDetailsForm
          {...card}
          onBack={() => setIsEditing(false)}
          onSubmit={() => console.log('test')}
        />
      ) : (
        <CardDetailsView {...card} onEdit={() => setIsEditing(true)} />
      )}
    </SPModal>
  )
}

type ViewProps = NCards.Card & { onEdit: VoidFunction }
function CardDetailsView(props: ViewProps) {
  const [areAnswersVisible, setAreAnswersVisible] = useState(false)

  return (
    <>
      <SPModalHeader data-testid="card-details-view">
        <div className="justify-space-between">
          <span>Card Details</span>
          <SPButton data-testid="card-details-edit-btn" onClick={props.onEdit}>
            Edit
          </SPButton>
        </div>
      </SPModalHeader>
      <SPModalContent className="flex-column align-center justify-center">
        <div className="flex-row-reverse">
          <Icon
            data-testid="view-answers-btn"
            name={!areAnswersVisible ? 'arrow down' : 'arrow up'}
            onClick={() => setAreAnswersVisible(!areAnswersVisible)}
          />
        </div>
        <List horizontal className="w-full" style={styles.flex}>
          <ListItem className="flex-1">
            <SPText className="w-full" value={props.question} />
          </ListItem>
          <ListItem className="flex-1">
            <List style={styles.p0}>
              {areAnswersVisible &&
                props.answers.map((answer, index) => {
                  return (
                    <ListItem key={answer.id} data-testid={`answers-${index}`}>
                      <SPText className="w-full" value={answer.content} />
                    </ListItem>
                  )
                })}
            </List>
          </ListItem>
        </List>
      </SPModalContent>
    </>
  )
}

const CardFormValidationSchema = Yup.object().shape({
  question: Yup.string().min(3).required('Required'),
  answers: Yup.array(
    Yup.object().shape({
      content: Yup.string().min(3).required('Required'),
    }),
  )
    .min(1)
    .required('Required'),
})
type FormProps = NCards.Card & { onBack: VoidFunction; onSubmit: VoidFunction }
function CardDetailsForm(props: FormProps) {
  const handleSubmit = (values: any) => {
    return
  }

  const form = useFormik({
    initialValues: {
      question: props.question,
      answers: props.answers,
    },
    validationSchema: CardFormValidationSchema,
    onSubmit: handleSubmit,
  })

  const handleAddAnswer = () => {
    const answers = [...form.values.answers, NAnswers.Initial({})]
    form.setFieldValue('answers', answers)
  }

  const handleDeleteAnswer = (id: string) => {
    const answers = form.values.answers.filter((answer) => answer.id !== id)
    form.setFieldValue('answers', answers)
  }

  const handleChangeAnswer = (id: string, newAnswerContent: string) => {
    const answers = form.values.answers.map((answer) =>
      answer.id === id ? { ...answer, content: newAnswerContent } : answer,
    )
    form.setFieldValue('answers', answers)
  }

  const handleChangeQuestion = (question: string) => {
    form.setFieldValue('question', question)
  }

  const getAnswerError = (index: number) => {
    const answerErrors = form.errors.answers || []
    const touched = !_.isEmpty(form.touched)
    return !!answerErrors[index] && touched
  }

  const getQuestionError = () => {
    const questionError = form.errors.question
    const touched = !_.isEmpty(form.touched)
    return questionError && touched
  }

  return (
    <>
      <SPModalHeader data-testid="card-details-form">
        <div>Edit Card</div>
        <SPButton
          data-testid="card-details-form-back-btn"
          onClick={props.onBack}
        />
      </SPModalHeader>
      <SPModalContent className="flex-column align-center justify-center">
        <Form className="w-full">
          <List horizontal className="flex" style={styles.flex}>
            <List.Item className="flex-1">
              <Form.Input
                name="question-content"
                placeholder="Enter question here"
                className="w-full"
                error={getQuestionError()}
                value={form.values.question}
                onChange={(e) => {
                  handleChangeQuestion(e.target.value)
                }}
              />
            </List.Item>
            <List.Item className="flex-1">
              <List style={styles.p0}>
                {form.values.answers.map((answer, index) => (
                  <List.Item
                    key={answer.id}
                    className="flex"
                    style={styles.flex}
                  >
                    <Form.Input
                      value={answer.content}
                      placeholder="Enter answer here"
                      className="w-full"
                      error={getAnswerError(index)}
                      name="answer-content"
                      onChange={(e) =>
                        handleChangeAnswer(answer.id, e.target.value)
                      }
                    />

                    <SPButton
                      size="small"
                      style={styles.bgWhite}
                      disabled={form.values.answers.length === 1}
                      icon={<Icon name="x" />}
                      onClick={() => handleDeleteAnswer(answer.id)}
                    />
                  </List.Item>
                ))}
                <List.Item style={styles.textAlignRight}>
                  <SPButton
                    size="small"
                    style={styles.bgWhite}
                    icon={<Icon name="plus" color="green" />}
                    onClick={handleAddAnswer}
                  />
                </List.Item>
              </List>
            </List.Item>
          </List>

          <div className="flex-row-reverse">
            <SPButton
              data-testid="card-details-form-save-btn"
              size="small"
              color="green"
              type="submit"
              onClick={form.submitForm}
            >
              Save
            </SPButton>
          </div>
        </Form>
      </SPModalContent>
    </>
  )
}
