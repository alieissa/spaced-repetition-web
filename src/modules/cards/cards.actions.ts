/** @format */

import { RequestError } from 'src/types'
import { Either } from 'src/utils/either'
import { NCards } from './cards.types'

export type CheckAnswer = {
   type: 'CheckAnswer'
   id: NCards.Card['id']
}

export type AnswerChecked = {
   type: 'AnswerChecked'
   result: Either<RequestError, NCards.State['check'][string]>
   id: NCards.Card['id']
}

export type UpdateQuality = {
   type: 'UpdateCardQuality',
   id: NCards.Card['id'],
}

export type QualityUpdated = {
   type: 'QualityUpdated'
   result: Either<RequestError, NCards.State['check'][string]>
   id: NCards.Card['id']
}

export type CardsAction =
  | CheckAnswer
  | AnswerChecked
  | UpdateQuality
  | QualityUpdated
