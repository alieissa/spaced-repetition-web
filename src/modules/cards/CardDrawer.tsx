/** @format */
import { MouseEventHandler } from 'react'
import 'semantic-ui-css/semantic.min.css'
import { Button, Form, Icon, Input, List } from 'semantic-ui-react'
import 'src/App.css'
import { NAnswers } from 'src/modules/answers'
import { styles } from 'src/styles'

export type CardFormProps = {
  id: string
  question: string
  answers: NAnswers.Answer[]
  onChangeQuestion: (question: string) => void
  onChangeAnswer: (id: string, content: string) => void
  onDeleteAnswer: (id: string) => void
  onAddAnswer: VoidFunction
  onSubmitForm?: any
  onCancel?: MouseEventHandler
}

export default function CardDrawer(props: CardFormProps) {
  return (
    <Form className="w-full">
      <List horizontal className="flex" style={styles.flex}>
        <Input
          name="question-content"
          placeholder="Enter question here"
          className="w-full"
          value={props.question}
        />

        <List.Item className="flex-1">
          <List style={styles.p0}>
            {props.answers.map((answer) => (
              <List.Item key={answer.id} className="flex" style={styles.flex}>
                {answer.content}
              </List.Item>
            ))}
            <List.Item style={styles.textAlignRight}>
              <Button
                size="small"
                style={styles.bgWhite}
                icon={<Icon name="plus" color="green" />}
              />
            </List.Item>
          </List>
        </List.Item>
      </List>
    </Form>
  )
}
