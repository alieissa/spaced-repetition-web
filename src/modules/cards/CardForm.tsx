/** @format */
import * as _ from 'lodash'
import { useState } from 'react'
import 'semantic-ui-css/semantic.min.css'
import { Form, Icon } from 'semantic-ui-react'
import 'src/App.css'
import { SPButtonIcon, SPInput, SPList, SPListItem } from 'src/components'
import { NAnswers } from 'src/modules/answers'
import { styles } from 'src/styles'

export type CardFormProps = {
  id: string
  question: string
  answers: NAnswers.Answer[]
  onChangeQuestion: (question: string) => void
  onChangeAnswer: (id: string, content: string) => void
  onDeleteAnswer: (id: string) => void
  onAddAnswer: VoidFunction
}

export default function CardForm(props: CardFormProps) {
  const [areAnswersVisible, setAreAnswersVisible] = useState(false)

  return (
    <Form className="w-full">
      <div className="flex-row-reverse">
        <Icon
          data-testid="view-answers-btn"
          name={!areAnswersVisible ? 'arrow down' : 'arrow up'}
          onClick={() => setAreAnswersVisible(!areAnswersVisible)}
        />
      </div>
      <SPList horizontal className="flex" style={styles.flex}>
        <SPListItem className="flex-1">
          <SPInput
            name="question-content"
            placeholder="Enter question here"
            className="w-full"
            value={props.question}
            onChange={(e) => {
              props.onChangeQuestion(e.target.value)
            }}
          />
        </SPListItem>
        <SPListItem className="flex-1">
          {areAnswersVisible && (
            <SPList style={styles.p0}>
              {props.answers.map((answer) => (
                <SPListItem
                  key={answer.id}
                  className="flex"
                  style={styles.flex}
                >
                  <SPInput
                    value={answer.content}
                    placeholder="Enter answer here"
                    className="w-full"
                    name="answer-content"
                    onChange={(e) => {
                      props.onChangeAnswer(answer.id, e.target.value || '')
                    }}
                  />

                  <SPButtonIcon
                    size="small"
                    style={styles.bgWhite}
                    disabled={_.size(props.answers) === 1}
                    icon="x"
                    onClick={() => {
                      props.onDeleteAnswer(answer.id)
                    }}
                  />
                </SPListItem>
              ))}
              <SPListItem style={styles.textAlignRight}>
                <SPButtonIcon
                  size="small"
                  style={styles.bgWhite}
                  icon="plus"
                  onClick={() => {
                    props.onAddAnswer()
                  }}
                />
              </SPListItem>
            </SPList>
          )}
        </SPListItem>
      </SPList>
    </Form>
  )
}
