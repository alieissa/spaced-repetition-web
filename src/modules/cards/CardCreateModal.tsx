/** @format */

import { useEffect } from 'react'
import { useNavigate, useParams } from 'react-router-dom'
import { Message } from 'semantic-ui-react'
import {
  SPButton,
  SPModal,
  SPModalActions,
  SPModalContent,
  SPModalHeader,
} from 'src/components'
import { async } from 'src/utils'
import CardForm from './CardForm'
import { useCardCreate, useCardForm } from './cards.hooks'
import { NCards } from './cards.types'

/**
 * This component contains the create card modal and the button that is used to toggle
 * the modal.
 */
export default function CardCreateModal() {
  const params = useParams()
  const navigate = useNavigate()
  const [createCardStatus, createCard] = useCardCreate(params.deckId!)

  const initCard = NCards.Initial({})

  const {
    form,
    getQuestionError,
    getAnswerError,
    handleAddAnswer,
    handleDeleteAnswer,
    handleChangeAnswer,
    handleChangeQuestion,
  } = useCardForm(initCard, () => {
    createCard(form.values)
  })

  const handleCloseModal = () => navigate(-1)

  useEffect(() => {
    if (createCardStatus.type === 'Success') {
      handleCloseModal()
    }
    //eslint-disable-next-line react-hooks/exhaustive-deps
  }, [createCardStatus.type])

  return (
    <SPModal
      data-testid="card-create-modal"
      open={true}
      onClose={handleCloseModal}
    >
      <SPModalHeader>Create Card</SPModalHeader>
      <SPModalContent className="flex-column align-center justify-center">
        {async.match(createCardStatus)({
          Untriggered: () => null,
          Success: () => null,
          Loading: () => null,
          Failure: () => (
            <Message
              data-testid="card-create-error"
              negative
              className="w-inherit"
            >
              <Message.Header>Error: Failed to create card</Message.Header>
            </Message>
          ),
        })}

        <CardForm
          {...form.values}
          areAnswersVisible={true}
          getQuestionError={getQuestionError}
          getAnswerError={getAnswerError}
          onAddAnswer={handleAddAnswer}
          onChangeAnswer={handleChangeAnswer}
          onDeleteAnswer={handleDeleteAnswer}
          onChangeQuestion={handleChangeQuestion}
        />
      </SPModalContent>
      <SPModalActions>
        <SPButton onClick={handleCloseModal}>Cancel</SPButton>
        <SPButton
          data-testid="card-create-save-btn"
          color="green"
          onClick={form.submitForm}
        >
          Save
        </SPButton>
      </SPModalActions>
    </SPModal>
  )
}
