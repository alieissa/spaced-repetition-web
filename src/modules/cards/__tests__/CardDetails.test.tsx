/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, screen } from '@testing-library/react'
import user from '@testing-library/user-event'
import { rest } from 'msw'
import { setupServer } from 'msw/lib/node'
import { renderWithProviders } from 'src/utils/test-utils'
import CardDetails from '../CardDetails'
import { NCards } from '../cards.types'

const deckId = faker.string.uuid()
const cardCreateUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks/${deckId}/cards`

export const handlers = [
  rest.post(cardCreateUrl, async (__, res, ctx) => {
    return res(ctx.status(200), ctx.delay(100))
  }),
]

const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('CardDetails', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(
        <CardDetails {...NCards.Initial({})} />,
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('interaction', () => {
    describe('CardDetailsView', () => {
      it('should display card form when clicking on edit button', async () => {
        // Assemble
        renderWithProviders(<CardDetails {...NCards.Initial({})} />)

        // Act
        const cardDetailsEditBtn = screen.getByTestId('card-details-edit-btn')
        await act(() => user.click(cardDetailsEditBtn))

        // Assert
        const cardDetailsEditForm = screen.queryByTestId('card-details-form')
        expect(cardDetailsEditForm).toBeInTheDocument()
      })

      it('should display answers of question when eye icon is clicked', async () => {
        // Assemble
        renderWithProviders(<CardDetails {...NCards.Initial({})} />)

        // Act
        const viewAnswersBtn = screen.getByTestId('view-answers-btn')
        await act(() => user.click(viewAnswersBtn))

        // Assert
        const answer = screen.queryByTestId('answers-0')
        expect(answer).toBeInTheDocument()
      })
    })
  })
})
