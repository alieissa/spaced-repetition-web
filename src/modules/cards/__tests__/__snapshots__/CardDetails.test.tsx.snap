// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDetails view should render correctly 1`] = `
<DocumentFragment>
  <button
    class="ui button"
    data-testid="card-details-edit-btn"
  />
  <div>
    <button
      class="ui button"
      data-testid="view-answers-btn"
    />
  </div>
</DocumentFragment>
`;
